### PR TITLE
Refactor MergeTreeDataSelectExecutor

### PR DESCRIPTION
**Pull request by KochetovNicolai**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

**Comment by KochetovNicolai** [comment]
```
2021-05-29 00:41:33 01149_zookeeper_mutation_stuck_after_replace_partition:                 [ FAIL ] 1.49 sec. - result differs with reference:
2021-05-29 00:41:33 --- /usr/share/clickhouse-test/queries/0_stateless/01149_zookeeper_mutation_stuck_after_replace_partition.reference	2021-05-28 22:05:56.000000000 +0300
2021-05-29 00:41:33 +++ /tmp/clickhouse-test/0_stateless/01149_zookeeper_mutation_stuck_after_replace_partition.stdout	2021-05-29 00:41:33.122548980 +0300
2021-05-29 00:41:33 @@ -7,6 +7,7 @@
2021-05-29 00:41:33  1	1
2021-05-29 00:41:33  2	2
2021-05-29 00:41:33  mt	0	0_1_1_0	2
2021-05-29 00:41:33 +rmt	0	0_0_0_0_1	2
2021-05-29 00:41:33  rmt	0	0_3_3_0	2
2021-05-29 00:41:33  0000000000	UPDATE s = concat(\'s\', toString(n)) WHERE 1	[]	0	1
2021-05-29 00:41:33  0000000001	DROP COLUMN s	[]	0	1
2021-05-29 00:41:33 
```

**Comment by KochetovNicolai** [comment]
```
2021-05-29 00:38:11 00910_zookeeper_test_alter_compression_codecs_long:                     [ FAIL ] 36.28 sec. - return code 159
2021-05-29 00:38:11 Received exception from server (version 21.7.1):
2021-05-29 00:38:11 Code: 159. DB::Exception: Received from localhost:9000. DB::Exception: Watching task /test/clickhouse/db/default/log/query-0000000004 is executing longer than distributed_ddl_task_timeout (=30) seconds. There are 2 unfinished hosts (2 of them are currently active), they are going to execute the query in background. 
2021-05-29 00:38:11 , result:
2021-05-29 00:38:11 
2021-05-29 00:38:11 
```

File: src/Processors/QueryPlan/ReadFromMergeTree.cpp [hunk 1]
```diff
@@ -140,16 +230,826 @@ Pipe ReadFromMergeTree::read()
...
+    const size_t min_marks_for_concurrent_read = MergeTreeDataSelectExecutor::minMarksForConcurrentRead(
+        settings.merge_tree_min_rows_for_concurrent_read,
+        settings.merge_tree_min_bytes_for_concurrent_read,
+        data_settings->index_granularity,
+        index_granularity_bytes,
+        sum_marks);
+
+    bool use_uncompressed_cache = settings.use_uncompressed_cache;
+    if (sum_marks > max_marks_to_use_cache)
+        use_uncompressed_cache = false;
```
Comment (CurtizJ) [inline src/Processors/QueryPlan/ReadFromMergeTree.cpp:275]: This quite huge part of code looks identical in 3 places. Is it possible to extract it to separate function?

**Review by CurtizJ — COMMENTED** [review]
In general looks ok.

File: src/Processors/QueryPlan/ReadFromMergeTree.cpp (hunk 1 above)
Comment (KochetovNicolai) [reply src/Processors/QueryPlan/ReadFromMergeTree.cpp:275]: ok, I will try

**Review by CurtizJ — APPROVED** [review]

**Comment by KochetovNicolai** [comment]
```
2021-06-01 19:53:41 00910_zookeeper_test_alter_compression_codecs_long:                     [ FAIL ] 35.47 sec. - return code 159
2021-06-01 19:53:41 Received exception from server (version 21.7.1):
2021-06-01 19:53:41 Code: 159. DB::Exception: Received from localhost:9000. DB::Exception: Watching task /test/clickhouse/db/default/log/query-0000000004 is executing longer than distributed_ddl_task_timeout (=30) seconds. There are 2 unfinished hosts (2 of them are currently active), they are going to execute the query in background. 
2021-06-01 19:53:41 , result:
2021-06-01 19:53:41 
```

**Pull request merged**